r/src/array_from_vector.cpp

```diff
@@ -868,6 +868,20 @@ std::shared_ptr<arrow::DataType> InferType(SEXP x) {
                                                      InferType(VECTOR_ELT(x, i)));
         }
         return std::make_shared<StructType>(std::move(fields));
+      } else {
+        R_xlen_t n = XLENGTH(x);
+        if (n == 0) {
+          Rcpp::stop("received length-0 list");
+        }
+        std::shared_ptr<arrow::DataType> element_type = InferType(VECTOR_ELT(x, 0));
+        for (R_xlen_t i = 1; i < n; i++) {
```

Comment:
I also see this util function `arrow::Status CheckCompatibleStruct`, would it make more sense to define `CheckCompatibleList` and do this across-row consistency check there?